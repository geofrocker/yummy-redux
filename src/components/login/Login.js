import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux'
import {bindActionCreators} from 'redux'
import * as LoginActions from '../../actions/loginActions'
import toastr from 'toastr'
import {Redirect} from 'react-router-dom'

class Login extends Component {
  constructor(props, context){
    super(props, context)
    this.state={
        userData:{username:'',password:''},
        message:'',
        redirect:false
    }
    
  }

  handleChange = (event) => {
    const field = event.target.name
    let userData = this.state.userData
    userData[field] = event.target.value
    this.setState({ userData })
  }
    
  loginUser = (e) =>{
    e.preventDefault();
    this.props.actions.login(this.state.userData)
    .then(() =>{
      toastr.success('You are now logged In')
      this.setState({redirect:true})
    })
    .catch(xhr=>{
      return this.setState(() =>{
        return {
          message:xhr.response?xhr.response.data.error:'An Error occurred, Try again',
          userData:{username:'', password:''}
          }
    })
  })}

  render(){
      const {redirect,userData, message}=this.state
      if(redirect){
        return <Redirect to="/" />
      }
      return (
          <div className="Login">
              <h1>Please sign in</h1>
              {message
                  ? <div className="alert alert-danger col-xs-11">{message}</div>
                  : <div></div> 
              }
              <div className="jumbotron col-xs-11">
                  <form method="POST" id="login-form" onSubmit={this.loginUser}>
                      <div className="form-group">
                          <input type="text" className="form-control" name="username" placeholder="Username" onChange={this.handleChange} value={userData.username} required/>
                      </div>
                  
                      <div className="form-group">
                          <input type="password" className="form-control" name="password" placeholder="Password" onChange={this.handleChange} value={userData.password} required/>
                      </div>
                      
                      <input type="submit" className="btn btn-primary pull-right" value="Login"/>

                      <center>
                          <p>or</p>
                          <p><a href="/register">Click here to register</a></p>
                      </center>
                  </form>
              </div>
              
          </div>
      );
  }
}
Login.contextType ={
  router:PropTypes.object
}
Login.propTypes = {
  message: PropTypes.string.isRequired,
}

function mapStateToProps(state, ownProps){
  return {
    message:state.auth.message
  }
}

function mapDispatchToProps(dispatch){
  return{
    actions: bindActionCreators(LoginActions, dispatch)
  }
}
export default connect(mapStateToProps, mapDispatchToProps) (Login);