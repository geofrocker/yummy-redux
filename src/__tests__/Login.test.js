import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Login } from '../components/login/Login';

const props = {
  login: () => Promise.resolve(),
};

describe('login component', () => {
  const wrapper = shallow(<Login {...props} />);
  const preventDefault = jest.fn();

  expect(toJson(wrapper)).toMatchSnapshot();
  it('renders the Login class', () => {
    expect(wrapper.find('.Login')).toHaveLength(1);
  });

  it('renders five div jsx elements', () => {
    expect(wrapper.find('div')).toHaveLength(5);
  });
  it('renders the sign in form and submit data', () => {
    wrapper.setState({ email: 'geofrocker2@gmail.com', password: 'psalms' });
    wrapper.find('#login-form').simulate('submit', { preventDefault });
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(preventDefault).toBeCalled();
  });

  it('render the two inputs and hides password', () => {
    expect(wrapper.find('#password')).toHaveLength(1);
    expect(wrapper.find('#username')).toHaveLength(1);
    wrapper.find('#username').simulate('change', { target: { value: 'psalms' } });
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
