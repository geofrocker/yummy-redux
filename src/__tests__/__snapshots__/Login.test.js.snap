// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div
  className="Login"
>
  <h1>
    Please sign in
  </h1>
  <div />
  <div
    className="jumbotron col-xs-11"
  >
    <form
      id="login-form"
      method="POST"
      onSubmit={[Function]}
    >
      <div
        className="form-group"
      >
        <input
          className="form-control"
          id="username"
          name="username"
          onChange={[Function]}
          placeholder="Username"
          required={true}
          type="text"
          value=""
        />
      </div>
      <div
        className="form-group"
      >
        <input
          className="form-control"
          id="password"
          name="password"
          onChange={[Function]}
          placeholder="Password"
          required={true}
          type="password"
          value=""
        />
      </div>
      <input
        className="btn btn-primary pull-right"
        type="submit"
        value="Login"
      />
      <center>
        <p>
          or
        </p>
        <p>
          <a
            href="/register"
          >
            Click here to register
          </a>
        </p>
      </center>
    </form>
  </div>
</div>
`;

exports[`login component render the two inputs and hides password 1`] = `
<Redirect
  push={false}
  to="dashboard"
/>
`;

exports[`login component renders the sign in form and submit data 1`] = `
<div
  className="Login"
>
  <h1>
    Please sign in
  </h1>
  <div />
  <div
    className="jumbotron col-xs-11"
  >
    <form
      id="login-form"
      method="POST"
      onSubmit={[Function]}
    >
      <div
        className="form-group"
      >
        <input
          className="form-control"
          id="username"
          name="username"
          onChange={[Function]}
          placeholder="Username"
          required={true}
          type="text"
          value=""
        />
      </div>
      <div
        className="form-group"
      >
        <input
          className="form-control"
          id="password"
          name="password"
          onChange={[Function]}
          placeholder="Password"
          required={true}
          type="password"
          value=""
        />
      </div>
      <input
        className="btn btn-primary pull-right"
        type="submit"
        value="Login"
      />
      <center>
        <p>
          or
        </p>
        <p>
          <a
            href="/register"
          >
            Click here to register
          </a>
        </p>
      </center>
    </form>
  </div>
</div>
`;
